### ISSUE TITLE
[ArrowDataSource Performance] Offset buffer can be pre-grown in Parquet ByteArray reader

### ISSUE THREAD

**Issue by zhztheplayer** [Intel-bigdata/arrow#67]
Following workaround can accelerate reading on `lineitem.parquet` by up to 20% (530ms to 420ms in my case, 4096 batchsize):

```
index 57404a883..828d09a2b 100644
--- a/cpp/src/arrow/array/builder_base.h
+++ b/cpp/src/arrow/array/builder_base.h
@@ -93,7 +93,7 @@ class ARROW_EXPORT ArrayBuilder {
     if (min_capacity <= current_capacity) return Status::OK();
 
     // leave growth factor up to BufferBuilder
-    auto new_capacity = BufferBuilder::GrowByFactor(current_capacity, min_capacity);
+    auto new_capacity = BufferBuilder::GrowByFactor(current_capacity, std::max(min_capacity, 4096L));
     return Resize(new_capacity);
   }
```

A final solution is needed for solving this since we cannot just hard code a batch size here.

**Comment by zhouyuan** [comment]
@zhztheplayer can we make this as configurable? e.g., passing some alloc_hint via configuraton/env variable

**Comment by zhztheplayer** [comment]
@zhouyuan Yes I think it's feasible to use the batch size environment variable we discussed about last week. If columnar plugin uses the same option to configure.

**Comment by zhztheplayer** [comment]
@zhouyuan Sorry previous solution has a problem: `std::max` should be used instead of `std::min`, it might affects your test result.

**Comment by zhztheplayer** [comment]
Resolved in 4d166c4dbd9c971e6f5daba42928e41ab8540274.
